### ISSUE TITLE
The Grand Refactor

### ISSUE THREAD

**Issue by kflynn** [datawire/ambassador#606]
tl;dr:
====

20180807 status
-----------------

Much work is in progress (see issues with the `tgr606` label), the rest is blocked on that work completing.

Summary
----------
We're going to refactor Ambassador so that we can get to V2 ADS while being able to sanely test it and more quickly develop new features.

Current State of the World
==========================

Ambassador's first commit was on 30 March 2017, so it's about 15 months old right now. Many, many things have changed during the course of its life, and the code is showing its age in some places (especially where we didn't correctly predict the direction we'd need to take it). We need to refactor the code to make life better in these areas, including:

- Ambassador doesn't cleanly separate Kubernetes, Envoy, and configuration handling.

    For example, the logic that handles loading TLS certificates from Kubernetes secrets should be isolated to the Kubernetes interface. Presently it's split between the configuration code and the Kubernetes code, and the Kubernetes code has to understand when to restart Envoy...

- Ambassador is currently limited to Envoy's V1 configuration with hot restart.

    Envoy's V2 configuration adds functionality and makes very fast configuration updates via ADS feasible. Ambassador has customers asking for these things.

- Ambassador's tests aren't very intelligent.

    With proper separation of concerns within Ambassador, we could be much smarter about which tests need to happen for which kinds of changes, permitting us to spend testing time where it would be most valuable. We can't do that right now, so we end up testing inefficiently. This has a few major effects:

   - Ambassador's end-to-end tests are very costly.

       The end-to-end tests actually spin up a Kubernetes cluster and run traffic through it. While we have to support this in some cases (it's necessary when developing new features), it's _very_ slow, and it's kind of pointless when we know that we haven't changed Envoy versions.

   - Ambassador doesn't run tests as often as it should.

       Since the end-to-end tests are so costly, we've disabled them for most builds. This is a problem, however, since:

   - Ambassador requires the end-to-end tests for full test coverage.

       Given that Ambassador doesn't correctly separate e.g. the Kubernetes interface from the configuration handler, the only way to correctly test that the configuration handler will correctly respond to a change from the Kubernetes interface is _actually to run Kubernetes and alter the Kubernetes configuration_.

   - Ambassador's diagnostics UI isn't tested in any automated way.

       The diagnostics _service_ is partly tested, but the web UI is not.

- Ambassador's IR implementation is needlessly complex.

    Ambassador functions by taking a set of Ambassador configuration resources (the _Ambassador config_), translating that to an intermediate representation (the _IR_), and then translating the IR to an Envoy configuration (the _Envoy config_). The IR forms the basis for Ambassador's logic around canaries, etc., and for Ambassador's diagnostics service. However, its implementation is unpleasant to work with, especially when adding new features.

Goals for the Refactor
======================

We're going to try to address these specific things in this refactor:

- Cleanly separate the various functional units within Ambassador so that we can write more intelligent, more easily extensible tests.
- Clean up the IR implementation to make it easier and faster to add new features.
- Switch Ambassador to the V2 Envoy config.
- Switch Ambassador to use ADS.

Broad Path Forward
==================

We've landed some tactical improvements around parallelizing the end-to-end tests. Next steps will be tracked as separate issues under this epic. Broadly speaking, though:

- Ambassador can learn to generate V2 configs in parallel with everything else.
   - The IR->V1 translation is driven with Jinja2; a similar IR->V2 translation shouldn't need to wait for anything else.
   - A critical piece here is being able to be sure that the V2 config generated from a given IR matches the V1 config for the same IR.

- Making the tests intelligent depends on separating the functional units properly.
   - We need the interfaces between units to allow tests to drive the units independently.

- Improving the IR implementation is probably also orthogonal to other tasks.
   - The IR is relatively better isolated than some other elements.
